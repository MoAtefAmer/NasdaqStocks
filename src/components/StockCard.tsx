import {Dimensions, StyleSheet} from 'react-native';
import {theme} from '../constants/theme';
import {Card, Text} from 'react-native-paper';
import ActiveChip from './ActiveChip';

interface StockCardProps {
  ticker: string;
  name: string;
  active: boolean;
  primaryExchange: string;
}

const StockCard = (props: StockCardProps) => {
  const {ticker, name, active, primaryExchange} = props;
  const screenWidth = Dimensions.get('window').width;
  const itemWidth = screenWidth / 2 - 20;
  return (
    <Card style={[styles.card, {width: itemWidth}]}>
      <Card.Title
        titleStyle={styles.titleStyle}
        subtitleStyle={styles.subtitleStyle}
        title={ticker}
        subtitle={primaryExchange}
        right={() => active && <ActiveChip />}
      />
      <Card.Content>
        <Text
          style={{color: 'white'}}
          variant="bodySmall"
          numberOfLines={3}
          ellipsizeMode="tail">
          {name}
        </Text>
      </Card.Content>
    </Card>
  );
};

export default StockCard;

const styles = StyleSheet.create({
  card: {
    padding: 10,
    backgroundColor: theme.colors.primary,
    margin: 10,
    marginBottom: 20,
    height: 150,
  },
  titleStyle: {
    color: 'white',
    fontSize: 20,
    fontWeight: '500',
  },
  subtitleStyle: {
    color: theme.colors.logoColor,
    fontSize: 12,
    fontWeight: '800',
  },
  text: {
    color: 'white',
    fontSize: 16,
  },
});
