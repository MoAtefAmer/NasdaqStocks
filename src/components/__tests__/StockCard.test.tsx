import React from 'react';
import { render } from '@testing-library/react-native';
import StockCard from '../StockCard';

describe('StockCard Component', () => {
  const mockProps = {
    ticker: 'AAPL',
    name: 'Apple Inc.',
    primaryExchange: 'NASDAQ',
  };

  it('renders correctly', () => {
    const { getByText } = render(<StockCard {...mockProps} />);

    expect(getByText('AAPL')).toBeTruthy();
    expect(getByText('NASDAQ')).toBeTruthy();
    expect(getByText('Apple Inc.')).toBeTruthy();
  });

  it('truncates long text with ellipsis', () => {
    const longName =
      'This is a very long company name that should be truncated in the StockCard component';
    const { getByText } = render(<StockCard {...mockProps} name={longName} />);

    const textComponent = getByText(longName);
    expect(textComponent.props.numberOfLines).toBe(3);
    expect(textComponent.props.ellipsizeMode).toBe('tail');
  });

  it('memoizes cardStyle to avoid unnecessary re-renders', () => {
    const { rerender, getByTestId } = render(<StockCard {...mockProps} />);
    const firstRenderCard = getByTestId('card-container-outer-layer');

    rerender(<StockCard {...mockProps} />); // Re-render with the same props
    const secondRenderCard = getByTestId('card-container-outer-layer');

    expect(firstRenderCard.props.style).toBe(secondRenderCard.props.style);
  });

  it('matches snapshot', () => {
    const tree = render(<StockCard {...mockProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
