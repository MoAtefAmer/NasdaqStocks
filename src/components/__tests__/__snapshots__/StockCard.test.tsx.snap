// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockCard Component matches snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgb(120, 69, 172)",
      "borderRadius": 12,
      "height": 150,
      "margin": 10,
      "marginBottom": 20,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
      "width": 355,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgb(120, 69, 172)",
        "borderRadius": 12,
        "elevation": 1,
        "flex": 1,
        "padding": 10,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 16,
            },
            {
              "minHeight": 72,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": 0.15,
                    "lineHeight": 24,
                  },
                  [
                    {
                      "minHeight": 30,
                      "paddingRight": 16,
                    },
                    {
                      "marginBottom": 0,
                    },
                    {
                      "color": "rgb(255, 255, 255)",
                      "fontSize": 20,
                      "fontWeight": "500",
                    },
                  ],
                ],
              ]
            }
          >
            AAPL
          </Text>
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "400",
                    "letterSpacing": 0.25,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginVertical": 0,
                      "minHeight": 20,
                      "paddingRight": 16,
                    },
                    {
                      "color": "#0088ce",
                      "fontSize": 12,
                      "fontWeight": "800",
                    },
                  ],
                ],
              ]
            }
          >
            NASDAQ
          </Text>
        </View>
        <View />
      </View>
      <View
        borderRadiusStyles={{}}
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingBottom": 16,
            },
            undefined,
          ]
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={3}
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "letterSpacing": 0.4,
                  "lineHeight": 16,
                },
                {
                  "color": "rgb(255, 255, 255)",
                },
              ],
            ]
          }
        >
          Apple Inc.
        </Text>
      </View>
    </View>
  </View>
</View>
`;
